torchvision/prototype/datasets/utils/_dataset.py

```diff
@@ -23,29 +21,56 @@ class DatasetType(enum.Enum):
     IMAGE = enum.auto()
 
 
+# TODO: this is just a dummy to not change the imports everywhere before the design is finalized
+class DatasetInfo:
+    pass
+
+
 class DatasetConfig(FrozenBunch):
     # This needs to be Frozen because we often pass configs as partial(func, config=config)
     # and partial() requires the parameters to be hashable.
     pass
 
 
-class DatasetInfo:
+SENTINEL = object()
+
+
+class DatasetOption:
```

Comment:
What is the reason for having a `DatasetOption` class instead of relying on standard Python parameter passing + docstrings?

I'm not sure we should try to automate input validation or docstring generation at all cost TBH. If it forces us to deviate from our usual Python development practices, it may not be worth it?

Reply:
Right now, we instantiate all `Dataset` objects at import time

https://github.com/pytorch/vision/blob/44e93beb0833ff51ffe75d994fe660434335b53d/torchvision/prototype/datasets/_api.py#L21-L23

At this stage no specific configuration is known. So something like

```py
class MyDataset(Dataset):
    def __init__(self, foo="foo", bar=1):
        if not isinstance(bar, int):
            raise TypeError
```

is not possible. 

Now, we can of course change the API to only instantiate the objects at runtime. But then we also need a different way to access the static data such as `categories` or the like, since they should be accessible without a specific configuration.

Reply:
> But then we also need a different way to access the static data such as categories or the like, since they should be accessible without a specific configuration.

Would it be reasonable to expect all parameters to have a default? This would allow us to instanciate the dataset objects whenever we need their static data.

Also, about the need for accessing meta-data without a config: perhaps we don't have that yet, but I could see a dataset that has different classes (or different "extra" field) depending on a parameter. Are we supporting this use-case?

Reply:
> Would it be reasonable to expect all parameters to have a default?

IMO, yes. We shouldn't have a dataset that is not loadable with `datasets.load("foo")`. Looking through our old datasets, I think there two different scenarios where this violated:

1. Datasets that cannot be automatically downloaded and need multiple files as input

    https://github.com/pytorch/vision/blob/44e93beb0833ff51ffe75d994fe660434335b53d/torchvision/datasets/coco.py#L28

2. Datasets with arguments that could very well have a default but don't for an probably unknown reason

    https://github.com/pytorch/vision/blob/44e93beb0833ff51ffe75d994fe660434335b53d/torchvision/datasets/mnist.py#L293

Since the download is handled differently and we no longer need to pass paths to the dataset, I currently don't see a reason to not assume it. If there ever is a case where we actually have no good default value, we can probably hack around it.

> Also, about the need for accessing meta-data without a config: perhaps we don't have that yet, but I could see a dataset that has different classes (or different "extra" field) depending on a parameter. Are we supporting this use-case?

Currently this is not supported. But I'm not sure what would be a good way to add this. With the design discussed in this thread, `datasets.info()` would give you the `Dataset` object instantiated with default parameters. Thus, we would always get the same meta data back. One way would be to accept `**options` there.

Reply:
Perhaps it would help to see how we would be documenting these new datasets, before we can make a decision on this.

I'm sceptical on general about not using normal parameter passing, defaults, and docs. Especially if we ever allow users to create their own classes, in we don't want them to learn yet another framework (but it's relevant for developers too). If it makes the docs much easier to write them we might have a reason, but for now it does not appear as obvious to me

Reply:
Let me write minimalistic docs to showcase.

Reply:
See https://1173984-73328905-gh.circle-artifacts.com/0/docs/datasets_prototype/index.html. Of course if we had more meta information filled in such as the citation, license, description, and so on, the pages would look not as empty.

Reply:
Thanks a lot @pmeier .

I have a few questions:

- how would we document the type of the parameters?
- how would we document extra methods on the datasets (if relevant)? What about extra attributes (again not sure if relevant)?
- how would we document additional info that is specific to each dataset? Stuff like "This dataset contains X images of cars and Y images of trucks".

Reply:
> how would we document the type of the parameters?

I would put another column in the table for the type. `DatasetOption` could have an `annotation` field for that. Let me add it to the examples.

> how would we document extra methods on the datasets (if relevant)?

IMO, we shouldn't have extra methods on the dataset objects. Currently, I don't think we have datasets that need this. Did you have anything in mind what other datasets might provide?

> What about extra attributes (again not sure if relevant)?

In our current stable API we sometimes add them to the docstring in the `Attributes:` section. We could simply have another table similar to `Options` that holds these. Let me add them to my examples.

> how would we document additional info that is specific to each dataset? Stuff like "This dataset contains X images of cars and Y images of trucks".

Given that this is very individual, I would simply have `description` field which takes a string input and puts it at the top of the generated documentation. Let me add an example.

Reply:
See https://1181665-73328905-gh.circle-artifacts.com/0/docs/datasets_prototype/index.html. I've taken the examples from https://paperswithcode.com/ for demo purposes.